[GFC] Add support for implicit grid tracks in the negative direction

yulunwu2 opened the pull request:
#### 4291549ea228120beabdd904cacb95428665d20f
<pre>
[GFC] Add support for implicit grid tracks in the negative direction
<a href="https://bugs.webkit.org/show_bug.cgi?id=300045">https://bugs.webkit.org/show_bug.cgi?id=300045</a>
&lt;<a href="https://rdar.apple.com/161843900">rdar://161843900</a>&gt;

Reviewed by Sammy Gill.

This PR updates GFC to handle grid items with negative positions by creating
implicit grid tracks in the negative direction. The implementation normalizes
the row and column numbers of explicitly placed grid items to their
coordinates implicit grid before performing layout.

This means all downstream code will work exclusively with non-negative
matrix indecies and eliminates the need for translation.

Combined changes:
* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::placeGridItems):
* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
(WebCore::Layout::ImplicitGrid::ImplicitGrid):
(WebCore::Layout::ImplicitGrid::insertUnplacedGridItem):
(WebCore::Layout::ImplicitGrid::gridAreas const):
(WebCore::Layout::ImplicitGrid::insertDefiniteRowItem):
* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.h:
(WebCore::Layout::ImplicitGrid::explicitGridRowStart const):
(WebCore::Layout::ImplicitGrid::explicitGridColumnStart const):
* Source/WebCore/layout/formattingContexts/grid/UnplacedGridItem.cpp:
(WebCore::Layout::UnplacedGridItem::explicitColumnStart const):
(WebCore::Layout::UnplacedGridItem::explicitColumnEnd const):
(WebCore::Layout::UnplacedGridItem::explicitRowStart const):
(WebCore::Layout::UnplacedGridItem::explicitRowEnd const):

Canonical link: <a href="https://commits.webkit.org/303275@main">https://commits.webkit.org/303275@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a2b9db78193ffb5e2a8452c78cb4c9687111e17

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131761 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4253 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42763 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139270 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83632 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ca76c891-cc3d-4dc1-bde5-de1777ee95ab) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133631 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4015 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100713 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134707 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3008 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118006 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81495 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2e2678c6-39a2-4382-89d6-7a9dec03ae7d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2895 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/737 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82492 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111664 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36133 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141916 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3922 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36710 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109083 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4003 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3528 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109245 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27698 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2981 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114287 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57132 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3976 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32697 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3808 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67423 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4068 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3936 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -39,62 +39,89 @@ namespace Layout {
+    int maximumRowIndex = gridTemplateRowsCount;
+    int maximumColumnIndex = gridTemplateColumnsCount;
+
+    auto processItem = [&](const UnplacedGridItem& item) {
```
IMO this name is a bit too vague since it doesn't really give any information on what this helper is doing

yulunwu2 replied:
updated to updateImplicitGridBounds

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.h:
```diff
@@ -39,11 +39,20 @@ struct GridAutoFlowOptions;
+    int explicitGridColumnStart() const { return m_explicitGridColumnStart; }
+
+    int translateRowToMatrixIndex(int explicitRowPosition) const { return explicitRowPosition + m_explicitGridRowStart; }
+    int translateColumnToMatrixIndex(int explicitColumnPosition) const { return explicitColumnPosition + m_explicitGridColumnStart; }
```
Do these really need to be part of the public API? Seems like they are really only things that are used internally by the `ImplicitGrid` so it might be confusing to expose them like this. Maybe they can just be static functions in the source file instead?

yulunwu2 replied:
Moved to be a private member function because of dependencies on m_ values.

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.h:
```diff
@@ -55,6 +64,8 @@ class ImplicitGrid {
     GridMatrix m_gridMatrix;
+    int m_explicitGridRowStart { 0 };
+    int m_explicitGridColumnStart { 0 };
 };
```
If I understand the spec correctly, we should be able to make these unsigned right? The location of the explicit grid within the implicit should be some non-negative value I think?

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -39,62 +39,89 @@ namespace Layout {
+    int implicitColumnsBeforeExplicitGrid = 0;
+    // Maximum indices start at the explicit grid size.
+    int maximumRowIndex = gridTemplateRowsCount;
+    int maximumColumnIndex = gridTemplateColumnsCount;
```
Is it possible for these to actually be unsigned? I know we compare it against the grid position, which can be negative, but if that is the case I think we take the absolute value anyways

yulunwu2 replied:
SGTM, refactoring to unsigned. Thanks for the review!

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -39,62 +39,90 @@ namespace Layout {
 // tracks and grow the grid.
-ImplicitGrid::ImplicitGrid(size_t gridTemplateColumnsCount, size_t gridTemplateRowsCount)
-    : m_gridMatrix(Vector(gridTemplateRowsCount, Vector<GridCell>(gridTemplateColumnsCount)))
+ImplicitGrid::ImplicitGrid(int gridTemplateColumnsCount, int gridTemplateRowsCount, const UnplacedGridItems& unplacedGridItems)
```
Can `gridTemplateColumnsCount` and `gridTemplateRowsCount` remain unsigned?

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -39,62 +39,90 @@ namespace Layout {
-    if (explicitColumnEnd <= explicitColumnStart) {
-        ASSERT_NOT_IMPLEMENTED_YET();
-        return;
-    }
```
Is this case actually completely handled yet? I see that after translating below you `ASSERT` to make sure that start < end, but how does this work if you have something like `grid-column-start: 3; grid-column-end: 1`?

yulunwu2 replied:
how handnormalizedColumnStartEnd and normalizedRowStartEnd now handle inverted ranges

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -39,62 +39,90 @@ namespace Layout {
-    auto rowsCount = static_cast<int>(this->rowsCount());
-    if (explicitRowStart > rowsCount || explicitRowEnd > rowsCount) {
+    // Grid items with inverted ranges (end <= start) are not yet supported.
+    if (columnEnd <= columnStart || rowEnd <= rowStart) {
```
Doesn't this contradict the `ASSERT`s above?

sammygill replied:
The `ASSERT_NOT_IMPLEMENTED` yet was making it clear that when checking the explicit positions from style we could not properly handle the scenario where end < start. Here I feel like it's a little less clear what state we are't properly supporting since it's happening after this translate that we're doing. Does that make sense?

yulunwu2 replied:
normalizedColumnStartEnd and normalizedRowStartEnd now handle inverted ranges

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -39,62 +39,90 @@ namespace Layout {
+    auto columnStart = translateColumnToMatrixIndex(explicitColumnStart);
+    auto columnEnd = translateColumnToMatrixIndex(explicitColumnEnd);
+    auto rowStart = translateRowToMatrixIndex(explicitRowStart);
+    auto rowEnd = translateRowToMatrixIndex(explicitRowEnd);
```
It's a bit difficult to understand what `MatrixIndex` really means when reading the code. Since I think we are translating from an explicit grid position to an implicit grid position the name should somehow include that?

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.h:
```diff
@@ -39,22 +39,33 @@ struct GridAutoFlowOptions;
+    // This is needed when items reference negative grid lines, requiring implicit tracks
+    // before line 1 of the explicit grid.
+    unsigned explicitGridRowStart() const { return m_explicitGridRowStart; }
+    unsigned explicitGridColumnStart() const { return m_explicitGridColumnStart; }
```
Does anyone currently use these? I didn't see it being called anywhere. If we need it do clients actually *need* to know where the explicit grid starts?

sammygill commented on Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
```diff
@@ -59,36 +59,86 @@ GridLayout::GridLayout(const GridFormattingContext& gridFormattingContext)
+        unplacedGridItems, gridTemplateColumnsTrackSizes.size(), gridTemplateRowsTrackSizes.size());
+
+    // Normalize all grid item positions by applying the offsets
+    UnplacedGridItems normalizedItems = unplacedGridItems;
```
Is there any reason we need to work on a copy rather than directly on `unplacedGridItems` or use a reference?

yulunwu2 replied:
I don't believe we need unplacedGridItems to be a const&. Removing const and skipping the copy.

sammygill commented on Source/WebCore/layout/formattingContexts/grid/UnplacedGridItem.cpp:
```diff
@@ -50,44 +51,60 @@ int UnplacedGridItem::explicitColumnStart() const
-    ASSERT_NOT_IMPLEMENTED_YET();
-    return { };
+    // Apply offset to normalize negative positions to non-negative matrix indices.
+    return position + static_cast<int>(m_columnNormalizationOffset);
```
What happens if someone calls these and the other similar functions before normalization happens? Should these functions just return the 0-indexed value or we actually need to normalize here? I suppose that begs the question when would we want to use this vs the normalized function

yulunwu2 replied:
You're right, this is the wrong place to do this. I've split out normalizedColumnStart and updated normalizedColumnStartEnd() to use the new methods.

sammygill commented on Source/WebCore/layout/formattingContexts/grid/UnplacedGridItem.h:
```diff
@@ -54,11 +55,18 @@ class UnplacedGridItem {
     int explicitRowStart() const;
     int explicitRowEnd() const;
```
We might want to consider making these private if nobody is using them anymore

yulunwu2 replied:
Done, thanks for the review!

sammygill commented on Source/WebCore/layout/formattingContexts/grid/UnplacedGridItem.h:
```diff
@@ -54,11 +55,18 @@ class UnplacedGridItem {
+    int normalizedColumnStart() const;
+    int normalizedColumnEnd() const;
+    int normalizedRowStart() const;
+    int normalizedRowEnd() const;
```
I'm wondering if these should return unsigned values? Can a normalized value be negative?

yulunwu2 committed:
[GFC] Add support for implicit grid tracks in the negative direction
https://bugs.webkit.org/show_bug.cgi?id=300045
<rdar://161843900>

Reviewed by Sammy Gill.

This PR updates GFC to handle grid items with negative positions by creating
implicit grid tracks in the negative direction. The implementation normalizes
the row and column numbers of explicitly placed grid items to their
coordinates implicit grid before performing layout.

This means all downstream code will work exclusively with non-negative
matrix indecies and eliminates the need for translation.

Combined changes:
* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::placeGridItems):
* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
(WebCore::Layout::ImplicitGrid::ImplicitGrid):
(WebCore::Layout::ImplicitGrid::insertUnplacedGridItem):
(WebCore::Layout::ImplicitGrid::gridAreas const):
(WebCore::Layout::ImplicitGrid::insertDefiniteRowItem):
* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.h:
(WebCore::Layout::ImplicitGrid::explicitGridRowStart const):
(WebCore::Layout::ImplicitGrid::explicitGridColumnStart const):
* Source/WebCore/layout/formattingContexts/grid/UnplacedGridItem.cpp:
(WebCore::Layout::UnplacedGridItem::explicitColumnStart const):
(WebCore::Layout::UnplacedGridItem::explicitColumnEnd const):
(WebCore::Layout::UnplacedGridItem::explicitRowStart const):
(WebCore::Layout::UnplacedGridItem::explicitRowEnd const):

Canonical link: https://commits.webkit.org/303275@main
